### PR TITLE
[Photo to Product] Show message when no text detected

### LINKED ISSUE #10269: Show a message when the image scanner fails to detect any text.

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #10269

## Description

Shows a message when no text is detected. This lets users know that they can try uploading a different image to auto-generate product details. 

## Testing instructions
- Since the feature is now under A/B test, please enable it by returning `true` in `AddProductFromImageEligibilityChecker`.
- Log in to a WPCom store and select the Products tab.
- Select the "+" button to add a new product.
- Select the manual option if the store has no product.
- Select Simple physical product in the product type action sheet.
- Notice the product details from image form is displayed.
- Select an image without any text in it.
- Observe that you see the message `No text detected. Please select another packaging photo or enter product details manually.` when no text is detected from the image.
- Try selecting a different image with text in it.
- Observe that the message is not displayed anymore.

## Screenshots

| Light | Dark |
|--------|--------|
| IMG_CC3D85FFE28D-1 | IMG_12E53840AB1D-1 | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift [hunk 1]
```diff
@@ -146,10 +151,17 @@ private extension AddProductFromImageViewModel {
...
                 analytics.track(event: .AddProductFromImage.scanCompleted(source: addProductSource, scannedTextCount: texts.count))
                 scannedTexts = texts.map { .init(text: $0, isSelected: true) }
                 [nameViewModel, descriptionViewModel].forEach { $0.reset() }
                 generateProductDetails()
+            } catch ScanError.noTextDetected {
+                shouldShowNoTextDetectedMessage = scannedTexts.isEmpty
+                DDLogError("⛔️ No text detected from image.")
             } catch {
                 analytics.track(event: .AddProductFromImage.scanFailed(source: addProductSource, error: error))
                 DDLogError("⛔️ Error scanning text from image: \(error)")
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift]: Should we also show the message here?

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift [hunk 2]
```diff
@@ -146,10 +151,17 @@ private extension AddProductFromImageViewModel {
         Task { @MainActor in
             do {
                 let texts = try await imageTextScanner.scanText(from: image)
+                guard texts.isNotEmpty else {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:160]: This case will not be tracked at all. Should we move the tracking to before this guard?

**Review by itsmeichigo — COMMENTED** [review]
Thanks for the update! I didn't realize that there would be a separator between the image and the message, which looks odd to me. How about moving this text to below `EditableImageView` in `AddProductFromImageFormImageView`?

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:160]: Good catch. Fixed in https://github.com/woocommerce/woocommerce-ios/pull/10282/commits/f4be8ebe0ab5eafe6a6e1a714dc39686e691cc4b

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift]: Started showing a different error message in https://github.com/woocommerce/woocommerce-ios/pull/10282/commits/f9dc80f70d7df365c78dd6d34a9dca4e2aee83f9

**Comment by selanthiraiyan** [comment]
> How about moving this text to below EditableImageView in AddProductFromImageFormImageView?

Done in https://github.com/woocommerce/woocommerce-ios/pull/10282/commits/7f773dae4e8898467e12389b2a7817bc7df403c7



---

Thanks for the review and valuable input, @itsmeichigo! The PR is ready for another look.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift [hunk 3]
```diff
@@ -147,18 +156,34 @@ private extension AddProductFromImageViewModel {
...
+
                 scannedTexts = texts.map { .init(text: $0, isSelected: true) }
                 [nameViewModel, descriptionViewModel].forEach { $0.reset() }
                 generateProductDetails()
             } catch {
-                analytics.track(event: .AddProductFromImage.scanFailed(source: addProductSource, error: error))
-                DDLogError("⛔️ Error scanning text from image: \(error)")
+                switch error {
+                case ScanError.noTextDetected:
+                    if scannedTexts.isEmpty {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: Should we remove this check and replace the `texts.isNotEmpty` check on line 160  with it?

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift [hunk 4]
```diff
@@ -147,18 +156,34 @@ private extension AddProductFromImageViewModel {
...
+                        textDetectionErrorMessage = Localization.noTextDetected
+                    }
+                    DDLogError("⛔️ No text detected from image.")
+                default:
+                    analytics.track(event: .AddProductFromImage.scanFailed(source: addProductSource, error: error))
+                    if scannedTexts.isEmpty {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:176]: I think this check redundant 🤔

**Review by itsmeichigo — APPROVED** [review]
Thanks for the update, the layout looks good to me! Leaving some questions regarding the `scannedText` check but the code looks good otherwise.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: This check ensures that we don't show an error when we already have a few scanned texts from a previous successful scan. (could be from a different image)

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:176]: https://github.com/woocommerce/woocommerce-ios/pull/10282#discussion_r1270497654

In the above comment, I explained why we have this check. Kindly let me know if I am missing something.

**Comment by selanthiraiyan** [comment]
@itsmeichigo I will merge this PR now to get this into release. Once you get time to check my replies, I will make changes in a future PR if needed. Thanks again for the review!

File: WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: @selanthiraiyan Sorry I don't understand your explanation. We are supposed to clear the scanned texts and error message from the previous image right? Why can it affect the next scan?

Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: @itsmeichigo Sorry, let me try to explain. Kindly let me know if you have any questions. 

> We are supposed to clear the scanned texts and error message from the previous image right? Why can it affect the next scan?

- The previous code before this implementation wasn't clearing the scanned texts from the previous image in case the new scan failed for some reason. 
- I hope it is because it is better to keep the `scannedTexts` of the previous image as it is probably of the same product. 
- i.e. The user took a photo from a different angle of the same product packaging to achieve better results or to add more items to the scanned texts list.

Kindly let me know your thoughts.

Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: I'd prefer to clear the scanned text after getting a new image to avoid weird edge cases. When the user takes a new picture, we expect a clean state, so everything should be reset.

Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageViewModel.swift:170]: Logged a subtask here #10293
